Generic Serializable causes compilation error with auto-value-parcel

I have a field with type `Class<T>`
`Class` is `Serializable` so `auto-value-parcel` tries to put that as a `Serializable`. This is fine. 

The problem is in the `CREATOR` method. Since it is a static context, it cannot access the type. Either you need to make `CREATOR` typed or you should remove the type inside `Creator`. This will cause a warning of course but that can be suppressed. 

Right now, it just does not compile. Here is a simple class to reproduce it. 

```java
@AutoValue
public abstract class Example<T extends Number> implements Parcelable {

  abstract Class<T> type();

}
```

A user here at Google has run into a similar problem. I was able to reproduce their case by modifying the `parameterizedType()` test method in `AutoValueParcelExtensionTest` like this:

```java
    JavaFileObject intf = JavaFileObjects.forSourceString("test.Intf", ""
            + "package test;\n"
            + "import android.os.Parcelable;\n"
            + "interface Intf extends Parcelable {}"
    );
    JavaFileObject source = JavaFileObjects.forSourceString("test.Test", ""
            + "package test;\n"
            + "import android.os.Parcelable;\n"
            + "import com.google.auto.value.AutoValue;\n"
            + "@AutoValue public abstract class Test<T extends Intf> implements Parcelable {\n"
            + "public abstract T tea();\n"
            + "}"
    );
...
```

Then the test fails like this:

<pre>
warning: No processor claimed any of these annotations: com.google.auto.value.AutoValue
warning: No processor claimed any of these annotations: javax.annotation.Generated
android/os/Parcel.java:22: warning: [rawtypes] found raw type: java.util.HashMap
  HashMap readHashMap(ClassLoader cl);
  ^
  missing type arguments for generic class java.util.HashMap&lt;K,V&gt;
android/os/Parcel.java:23: warning: [rawtypes] found raw type: java.util.ArrayList
  ArrayList readArrayList(ClassLoader cl);
  ^
  missing type arguments for generic class java.util.ArrayList&lt;E&gt;
android/os/Parcel.java:41: warning: [rawtypes] found raw type: java.util.Map
  void writeMap(Map in);
                ^
  missing type arguments for generic class java.util.Map&lt;K,V&gt;
android/os/Parcel.java:42: warning: [rawtypes] found raw type: java.util.List
  void writeList(List in);
                 ^
  missing type arguments for generic class java.util.List&lt;E&gt;
/SOURCE_OUTPUT/test/AutoValue_Test.java:10: warning: [rawtypes] found raw type: test.AutoValue_Test
  public static final Parcelable.Creator&lt;AutoValue_Test&gt; CREATOR = new Parcelable.Creator&lt;AutoValue_Test&gt;() {
                                         ^
  missing type arguments for generic class test.AutoValue_Test&lt;T&gt;
/SOURCE_OUTPUT/test/AutoValue_Test.java:10: warning: [rawtypes] found raw type: test.AutoValue_Test
  public static final Parcelable.Creator&lt;AutoValue_Test&gt; CREATOR = new Parcelable.Creator&lt;AutoValue_Test&gt;() {
                                                                                          ^
  missing type arguments for generic class test.AutoValue_Test&lt;T&gt;
/SOURCE_OUTPUT/test/AutoValue_Test.java:12: warning: [rawtypes] found raw type: test.AutoValue_Test
    public AutoValue_Test createFromParcel(Parcel in) {
           ^
  missing type arguments for generic class test.AutoValue_Test&lt;T&gt;
/SOURCE_OUTPUT/test/AutoValue_Test.java:13: warning: [rawtypes] found raw type: test.AutoValue_Test
      return new AutoValue_Test(
                 ^
  missing type arguments for generic class test.AutoValue_Test&lt;T&gt;
/SOURCE_OUTPUT/test/AutoValue_Test.java:14: error: non-static type variable T cannot be referenced from a static context
          (T) in.readParcelable(Test.class.getClassLoader())
           ^
/SOURCE_OUTPUT/test/AutoValue_Test.java:18: warning: [rawtypes] found raw type: test.AutoValue_Test
    public AutoValue_Test[] newArray(int size) {
           ^
  missing type arguments for generic class test.AutoValue_Test&lt;T&gt;
/SOURCE_OUTPUT/test/AutoValue_Test.java:19: warning: [rawtypes] found raw type: test.AutoValue_Test
      return new AutoValue_Test[size];
                 ^
  missing type arguments for generic class test.AutoValue_Test&lt;T&gt;
</pre>

----

There's a lot of noise from the `rawtypes` warnings, but there is one error in the bunch:

```
/SOURCE_OUTPUT/test/AutoValue_Test.java:14: error: non-static type variable T cannot be referenced from a static context
          (T) in.readParcelable(Test.class.getClassLoader())
           ^
```

The problem is that the static `CREATOR` field looks like this:

```java
final class AutoValue_Test<T extends Intf> extends $AutoValue_Test<T> {
  public static final Parcelable.Creator<AutoValue_Test> CREATOR = new Parcelable.Creator<AutoValue_Test>() {
    @Override
    public AutoValue_Test createFromParcel(Parcel in) {
      return new AutoValue_Test(
          (T) in.readParcelable(Test.class.getClassLoader())
      );
    }
    @Override
    public AutoValue_Test[] newArray(int size) {
      return new AutoValue_Test[size];
    }
  };
```

That cast to `(T)` fails because there is no `T` in scope (we are inside the initializer of a static field). In the actual user code, there was a second field of type `Optional<T>` which caused a second error because of the cast to `(Optional<T>)`. I mention that because a fix that detects just the exact `T` case would not be enough.

With plain `<T extends Parcelable>`, as in the unmodified `parameterizedType()` test, there is no cast in the generated code. The relevant code is [here](https://github.com/rharter/auto-value-parcel/blob/5bcad51aef5692487d6c0c9fac4c0ec5547c7be0/auto-value-parcel/src/main/java/com/ryanharter/auto/value/parcel/Parcelables.java#L181):

```java
      TypeName check = property.type instanceof TypeVariableName
          ? ((TypeVariableName) property.type).bounds.get(0)
          : property.type;
      if (!check.equals(PARCELABLE)) {
        block.add("($T) ", property.type);
      }
      block.add("in.readParcelable($T.class.getClassLoader())", autoValueType);
```

In the exact case of `<T extends Parcelable>`, this omits the cast, but in every other case it is inserted. A cast is in fact needed for the `Optional<T>` case, though it would need to be `(Optional)` or `(Optional<?>)`. Perhaps in the presence of a type parameter, the code should look to see if that parameter is mentioned in the type in question, and if so cast to the erased type?

--- a/auto-value-parcel/src/main/java/com/ryanharter/auto/value/parcel/AutoValueParcelExtension.java
+++ b/auto-value-parcel/src/main/java/com/ryanharter/auto/value/parcel/AutoValueParcelExtension.java
@@ -25,6 +25,7 @@ import com.squareup.javapoet.ParameterizedTypeName;
 import com.squareup.javapoet.TypeName;
 import com.squareup.javapoet.TypeSpec;
 import com.squareup.javapoet.TypeVariableName;
+import com.squareup.javapoet.WildcardTypeName;
 
 import java.util.ArrayList;
 import java.util.LinkedHashMap;
@@ -42,7 +43,6 @@ import javax.lang.model.element.TypeElement;
 import javax.lang.model.element.TypeParameterElement;
 import javax.lang.model.element.VariableElement;
 import javax.lang.model.type.ArrayType;
-import javax.lang.model.type.MirroredTypeException;
 import javax.lang.model.type.TypeKind;
 import javax.lang.model.type.TypeMirror;
 import javax.lang.model.type.TypeVariable;
@@ -230,7 +230,9 @@ public final class AutoValueParcelExtension extends AutoValueExtension {
       }
     }
 
-    subclass.addField(generateCreator(env, autoValueType, properties, type, typeAdapters));
+    List<? extends javax.lang.model.element.TypeParameterElement> typeParameters =
+        context.autoValueClass().getTypeParameters();
+    subclass.addField(generateCreator(env, autoValueType, properties, type, typeAdapters, typeParameters));
 
     ClassName superClass = ClassName.get(context.packageName(), classToExtend);
     List<? extends TypeParameterElement> tpes = context.autoValueClass().getTypeParameters();
@@ -374,13 +376,33 @@ public final class AutoValueParcelExtension extends AutoValueExtension {
   }
 
   FieldSpec generateCreator(ProcessingEnvironment env, TypeName autoValueType,
-      List<Property> properties, TypeName type, Map<TypeMirror, FieldSpec> typeAdapters) {
+      List<Property> properties, TypeName type, Map<TypeMirror, FieldSpec> typeAdapters,
+      List<? extends javax.lang.model.element.TypeParameterElement> typeParameters) {
+
     ClassName creator = ClassName.bestGuess("android.os.Parcelable.Creator");
-    TypeName creatorOfClass = ParameterizedTypeName.get(creator, type);
+    TypeName typeWithParameters;
+    if (!typeParameters.isEmpty()) {
+      TypeName[] elements = typeParameters.stream().map(
+              param -> {
+                TypeMirror bound = param.getBounds().get(0);
+                TypeName typeName = TypeName.get(bound);
+                // TODO: support types with multiple bounds.
+                return WildcardTypeName.subtypeOf(typeName);
+              }
+      ).toArray(TypeName[]::new);
+     typeWithParameters = ParameterizedTypeName.get((ClassName) type, elements);
+    } else {
+      typeWithParameters = type;
+    }
+    TypeName creatorOfClass = ParameterizedTypeName.get(creator, typeWithParameters);
 
     Types typeUtils = env.getTypeUtils();
     CodeBlock.Builder ctorCall = CodeBlock.builder();
-    ctorCall.add("return new $T(\n", type);
+    if (!typeParameters.isEmpty()) {
+      ctorCall.add("return ($T) new $T(\n", typeWithParameters, type);
+    } else {
+      ctorCall.add("return new $T(\n", type);
+    }
     ctorCall.indent().indent();
     boolean requiresSuppressWarnings = false;
     for (int i = 0, n = properties.size(); i < n; i++) {
@@ -407,7 +429,7 @@ public final class AutoValueParcelExtension extends AutoValueExtension {
     }
     createFromParcel
         .addModifiers(PUBLIC)
-        .returns(type)
+        .returns(typeWithParameters)
         .addParameter(ClassName.bestGuess("android.os.Parcel"), "in");
     createFromParcel.addCode(ctorCall.build());
 
@@ -418,7 +440,7 @@ public final class AutoValueParcelExtension extends AutoValueExtension {
         .addMethod(MethodSpec.methodBuilder("newArray")
             .addAnnotation(Override.class)
             .addModifiers(PUBLIC)
-            .returns(ArrayTypeName.of(type))
+            .returns(ArrayTypeName.of(typeWithParameters))
             .addParameter(int.class, "size")
             .addStatement("return new $T[size]", type)
             .build())

--- a/auto-value-parcel/src/main/java/com/ryanharter/auto/value/parcel/AutoValueParcelExtension.java
+++ b/auto-value-parcel/src/main/java/com/ryanharter/auto/value/parcel/AutoValueParcelExtension.java
@@ -210,7 +210,7 @@ public final class AutoValueParcelExtension extends AutoValueExtension {
 
     ImmutableMap<TypeMirror, FieldSpec> typeAdapters = getTypeAdapters(properties);
 
-    TypeName type = ClassName.get(context.packageName(), className);
+    ClassName type = ClassName.get(context.packageName(), className);
     TypeSpec.Builder subclass = TypeSpec.classBuilder(className)
         .addModifiers(FINAL)
         .addMethod(generateConstructor(properties))
@@ -230,7 +230,7 @@ public final class AutoValueParcelExtension extends AutoValueExtension {
       }
     }
 
-    List<? extends javax.lang.model.element.TypeParameterElement> typeParameters =
+    List<? extends TypeParameterElement> typeParameters =
         context.autoValueClass().getTypeParameters();
     subclass.addField(generateCreator(env, autoValueType, properties, type, typeAdapters, typeParameters));
 
@@ -376,8 +376,8 @@ public final class AutoValueParcelExtension extends AutoValueExtension {
   }
 
   FieldSpec generateCreator(ProcessingEnvironment env, TypeName autoValueType,
-      List<Property> properties, TypeName type, Map<TypeMirror, FieldSpec> typeAdapters,
-      List<? extends javax.lang.model.element.TypeParameterElement> typeParameters) {
+      List<Property> properties, ClassName type, Map<TypeMirror, FieldSpec> typeAdapters,
+      List<? extends TypeParameterElement> typeParameters) {
 
     ClassName creator = ClassName.bestGuess("android.os.Parcelable.Creator");
     TypeName typeWithParameters;
@@ -390,7 +390,7 @@ public final class AutoValueParcelExtension extends AutoValueExtension {
                 return WildcardTypeName.subtypeOf(typeName);
               }
       ).toArray(TypeName[]::new);
-     typeWithParameters = ParameterizedTypeName.get((ClassName) type, elements);
+     typeWithParameters = ParameterizedTypeName.get(type, elements);
     } else {
       typeWithParameters = type;
     }

--- a/auto-value-parcel/src/main/java/com/ryanharter/auto/value/parcel/Parcelables.java
+++ b/auto-value-parcel/src/main/java/com/ryanharter/auto/value/parcel/Parcelables.java
@@ -10,6 +10,7 @@ import com.squareup.javapoet.ParameterSpec;
 import com.squareup.javapoet.ParameterizedTypeName;
 import com.squareup.javapoet.TypeName;
 import com.squareup.javapoet.TypeVariableName;
+import com.squareup.javapoet.WildcardTypeName;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Set;
@@ -152,6 +153,27 @@ final class Parcelables {
     return false;
   }
 
+  private static void addTypecast(CodeBlock.Builder block, Property property){
+    TypeName typeToCastTo = property.type instanceof TypeVariableName
+            ? ((TypeVariableName) property.type).bounds.get(0)
+            : property.type;
+    if (property.type instanceof ParameterizedTypeName) {
+      ParameterizedTypeName param = (ParameterizedTypeName) property.type;
+      TypeName[] wildcardTypes = param.typeArguments.stream()
+              .map(typeArgument -> {
+                // TODO: support types with multiple bounds.
+                if (typeArgument instanceof  TypeVariableName) {
+                  TypeName bound = ((TypeVariableName) typeArgument).bounds.get(0);
+                  return WildcardTypeName.subtypeOf(bound);
+                }
+                return typeArgument;
+              }).toArray(TypeName[]::new);
+
+      typeToCastTo = ParameterizedTypeName.get(param.rawType, wildcardTypes);
+    }
+    block.add("($T) ", typeToCastTo);
+  }
+
   static void readValue(Types typeUtils, CodeBlock.Builder block, Property property,
       final TypeName parcelableType, TypeName autoValueType) {
     boolean needsNullCheck = needsNullCheck(typeUtils, property, parcelableType);
@@ -182,7 +204,7 @@ final class Parcelables {
           ? ((TypeVariableName) property.type).bounds.get(0)
           : property.type;
       if (!check.equals(PARCELABLE)) {
-        block.add("($T) ", property.type);
+        addTypecast(block, property);
       }
       block.add("in.readParcelable($T.class.getClassLoader())", autoValueType);
     } else if (parcelableType.equals(CHARSEQUENCE)) {
@@ -202,11 +224,10 @@ final class Parcelables {
     } else if (parcelableType.equals(STRINGARRAY)) {
       block.add("in.readStringArray()");
     } else if (parcelableType.equals(IBINDER)) {
-      if (property.type.equals(IBINDER)) {
-        block.add("in.readStrongBinder()");
-      } else {
-        block.add("($T) in.readStrongBinder()", property.type);
+      if (!property.type.equals(IBINDER)) {
+        addTypecast(block, property);
       }
+      block.add("in.readStrongBinder()");
     } else if (parcelableType.equals(OBJECTARRAY)) {
       block.add("in.readArray($T.class.getClassLoader())", autoValueType);
     } else if (parcelableType.equals(INTARRAY)) {
@@ -214,11 +235,10 @@ final class Parcelables {
     } else if (parcelableType.equals(LONGARRAY)) {
       block.add("in.createLongArray()");
     } else if (parcelableType.equals(SERIALIZABLE)) {
-      if (property.type.equals(SERIALIZABLE)) {
-        block.add("in.readSerializable()");
-      } else {
-        block.add("($T) in.readSerializable()", property.type);
+      if (!property.type.equals(SERIALIZABLE)) {
+        addTypecast(block, property);
       }
+      block.add("in.readSerializable()");
     } else if (parcelableType.equals(PARCELABLEARRAY)) {
       ArrayTypeName atype = (ArrayTypeName) property.type;
       if (!atype.componentType.equals(PARCELABLE)) {

--- a/auto-value-parcel/src/test/java/com/ryanharter/auto/value/parcel/AutoValueParcelExtensionTest.java
+++ b/auto-value-parcel/src/test/java/com/ryanharter/auto/value/parcel/AutoValueParcelExtensionTest.java
@@ -1899,15 +1899,15 @@ public class AutoValueParcelExtensionTest {
             "\n" +
             "@Generated(\"com.ryanharter.auto.value.parcel.AutoValueParcelExtension\")" +
             "final class AutoValue_Test<T extends Parcelable> extends $AutoValue_Test<T> {\n" +
-            "  public static final Parcelable.Creator<AutoValue_Test> CREATOR = new Parcelable.Creator<AutoValue_Test>() {\n" +
+            "  public static final Parcelable.Creator<AutoValue_Test<? extends Parcelable>> CREATOR = new Parcelable.Creator<AutoValue_Test<? extends Parcelable>>() {\n" +
             "    @Override\n" +
-            "    public AutoValue_Test createFromParcel(Parcel in) {\n" +
-            "      return new AutoValue_Test(\n" +
+            "    public AutoValue_Test<? extends Parcelable> createFromParcel(Parcel in) {\n" +
+            "      return (AutoValue_Test<? extends Parcelable>) new AutoValue_Test(\n" +
             "          in.readParcelable(Test.class.getClassLoader())\n" +
             "      );\n" +
             "    }\n" +
             "    @Override\n" +
-            "    public AutoValue_Test[] newArray(int size) {\n" +
+            "    public AutoValue_Test<? extends Parcelable>[] newArray(int size) {\n" +
             "      return new AutoValue_Test[size];\n" +
             "    }\n" +
             "  };\n" +
@@ -1919,6 +1919,82 @@ public class AutoValueParcelExtensionTest {
             "  @Override\n" +
             "  public void writeToParcel(Parcel dest, int flags) {\n" +
             "    dest.writeParcelable(tea(), flags);\n" +
+            "  }\n" +
+            "\n" +
+            "  @Override\n" +
+            "  public int describeContents() {\n" +
+            "    return 0;\n" +
+            "  }\n" +
+            "}");
+
+    assertAbout(javaSources())
+            .that(Arrays.asList(parcel, parcelable, source))
+            .processedWith(new AutoValueProcessor(ImmutableList.of(new AutoValueParcelExtension())))
+            .compilesWithoutError()
+            .and()
+            .generatesSources(expected);
+  }
+
+  @Test
+  public void parameterizedType_addsTypeCastForUnknownCompileTimeTypes() {
+    JavaFileObject source = JavaFileObjects.forSourceString("test.Test", ""
+            + "package test;\n"
+            + "import android.os.IBinder;\n"
+            + "import android.os.Parcelable;\n"
+            + "import com.google.auto.value.AutoValue;\n"
+            + "import java.io.Serializable;\n"
+            + "@AutoValue public abstract class Test<T extends String, U extends System> implements Parcelable {\n"
+            + "  interface ParcelableFoo<T, U> extends Parcelable{}\n"
+            + "  interface SerializableBar<T, U> extends Serializable{}\n"
+            + "  interface BinderBaz<T> extends IBinder{}\n"
+            + "\n"
+            + "  public abstract ParcelableFoo<T, U> foo();\n"
+            + "  public abstract SerializableBar<U, T> bar();\n"
+            + "  public abstract BinderBaz<T> baz();\n"
+            + "}"
+    );
+
+    JavaFileObject expected = JavaFileObjects.forSourceString("test/AutoValue_Test", ""
+            + "package test;\n" +
+            "\n" +
+            "import android.os.Parcel;\n" +
+            "import android.os.Parcelable;\n" +
+            "import java.lang.Override;\n" +
+            "import java.lang.String;\n" +
+            "import java.lang.SuppressWarnings;\n"+
+            "import java.lang.System;\n" +
+            "import javax.annotation.Generated;\n" +
+            "\n" +
+            "@Generated(\"com.ryanharter.auto.value.parcel.AutoValueParcelExtension\")" +
+            "final class AutoValue_Test<T extends String, U extends System> extends $AutoValue_Test<T, U> {\n" +
+            "  public static final Parcelable.Creator<AutoValue_Test<? extends String, ? extends System>> CREATOR = new Parcelable.Creator<AutoValue_Test<? extends String, ? extends System>>() {\n" +
+            "    @Override\n" +
+            "    @SuppressWarnings(\"unchecked\")\n" +
+            "    public AutoValue_Test<? extends String, ? extends System> createFromParcel(Parcel in) {\n" +
+            "      return (AutoValue_Test<? extends String, ? extends System>) new AutoValue_Test(\n" +
+            "          (Test.ParcelableFoo<? extends String, ? extends System>) in.readParcelable(Test.class.getClassLoader()),\n" +
+            "          (Test.SerializableBar<? extends System, ? extends String>) in.readSerializable(),\n" +
+            "          (Test.BinderBaz<? extends String>) in.readStrongBinder()\n" +
+            "      );\n" +
+            "    }\n" +
+            "    @Override\n" +
+            "    public AutoValue_Test<? extends String, ? extends System>[] newArray(int size) {\n" +
+            "      return new AutoValue_Test[size];\n" +
+            "    }\n" +
+            "  };\n" +
+            "\n" +
+            "  AutoValue_Test(\n"+
+            "    Test.ParcelableFoo<T, U> foo,\n" +
+            "    Test.SerializableBar<U, T> bar,\n" +
+            "    Test.BinderBaz<T> baz) {\n" +
+            "    super(foo, bar, baz);\n" +
+            "  }\n" +
+            "\n" +
+            "  @Override\n" +
+            "  public void writeToParcel(Parcel dest, int flags) {\n" +
+            "    dest.writeParcelable(foo(), flags);\n" +
+            "    dest.writeSerializable(bar());\n" +
+            "    dest.writeStrongBinder(baz());\n" +
             "  }\n" +
             "\n" +
             "  @Override\n" +
@@ -1960,17 +2036,17 @@ public class AutoValueParcelExtensionTest {
             "\n" +
             "@Generated(\"com.ryanharter.auto.value.parcel.AutoValueParcelExtension\")" +
             "final class AutoValue_Test<T extends Parcelable> extends $AutoValue_Test<T> {\n" +
-            "  public static final Parcelable.Creator<AutoValue_Test> CREATOR = new Parcelable.Creator<AutoValue_Test>() {\n" +
+            "  public static final Parcelable.Creator<AutoValue_Test<? extends Parcelable>> CREATOR = new Parcelable.Creator<AutoValue_Test<? extends Parcelable>>() {\n" +
             "    @Override\n" +
-            "    public AutoValue_Test createFromParcel(Parcel in) {\n" +
-            "      return new AutoValue_Test(\n" +
+            "    public AutoValue_Test<? extends Parcelable> createFromParcel(Parcel in) {\n" +
+            "      return (AutoValue_Test<? extends Parcelable>) new AutoValue_Test(\n" +
             "          in.readParcelable(Test.class.getClassLoader()),\n" +
             "          in.readString(),\n" +
             "          in.readInt()\n" +
             "      );\n" +
             "    }\n" +
             "    @Override\n" +
-            "    public AutoValue_Test[] newArray(int size) {\n" +
+            "    public AutoValue_Test<? extends Parcelable>[] newArray(int size) {\n" +
             "      return new AutoValue_Test[size];\n" +
             "    }\n" +
             "  };\n" +
@@ -2025,17 +2101,17 @@ public class AutoValueParcelExtensionTest {
         "\n" +
             "@Generated(\"com.ryanharter.auto.value.parcel.AutoValueParcelExtension\")" +
         "final class AutoValue_Test<T extends String> extends $AutoValue_Test<T> {\n" +
-        "  public static final Parcelable.Creator<AutoValue_Test> CREATOR = new Parcelable.Creator<AutoValue_Test>() {\n" +
+        "  public static final Parcelable.Creator<AutoValue_Test<? extends String>> CREATOR = new Parcelable.Creator<AutoValue_Test<? extends String>>() {\n" +
         "    @Override\n" +
-        "    public AutoValue_Test createFromParcel(Parcel in) {\n" +
-        "      return new AutoValue_Test(\n" +
+        "    public AutoValue_Test<? extends String> createFromParcel(Parcel in) {\n" +
+        "      return (AutoValue_Test<? extends String>) new AutoValue_Test(\n" +
         "          in.readString(),\n" +
         "          in.readString(),\n" +
         "          in.readInt()\n" +
         "      );\n" +
         "    }\n" +
         "    @Override\n" +
-        "    public AutoValue_Test[] newArray(int size) {\n" +
+        "    public AutoValue_Test<? extends String>[] newArray(int size) {\n" +
         "      return new AutoValue_Test[size];\n" +
         "    }\n" +
         "  };\n" +

